test/gtest/ucp/test_ucp_am.cc

```diff
@@ -799,6 +800,192 @@ UCS_TEST_P(test_ucp_am_nbx, rx_am_mpools,
+        ucs::fill_random(m_hdr_copy);
+        m_hdr = m_hdr_copy;
+        ucs_status_ptr_t pending_sptr = NULL;
+        req_counter = 0;
+        cb_counter = 0;
+
+        ucp::data_type_desc_t sdt_desc(m_dt, sbuf.ptr(), size);                               
+        ucp_mem_h memh;
+
+        if (prereg()) {
+            memh = sender().mem_map(sbuf.ptr(), size);
+        }
+
+        set_am_data_handler(receiver(), TEST_AM_NBX_ID, am_data_cb, this,
+                                    data_cb_flags);
+
+        pending_sptr = fill_sq(sdt_desc, memh, header_size, get_send_flag() | flags, data_cb_flags);
+
+        if (pending_sptr == send_completed) {
+            UCS_TEST_MESSAGE <<"Failed to get pending request";
+        }
+
+        ucs::fill_random(m_hdr_copy);
+
+        while(progress());
+        wait_for_value(&cb_counter, req_counter);
+        request_wait(pending_sptr);
+        EXPECT_EQ(req_counter, cb_counter);
+
+        if (prereg()) {
+            sender().mem_unmap(memh);
+        }
+    }
+
+    void check_rndv(size_t size, size_t header_size = 0ul,
+                           unsigned flags = 0, unsigned data_cb_flags = 0) {
+        mem_buffer sbuf(size, tx_memtype());
+        sbuf.pattern_fill(SEED);
+        m_hdr_copy.resize(header_size*2);
+        ucs::fill_random(m_hdr_copy);
+        m_hdr = m_hdr_copy;
+        ucs_status_ptr_t rndv_pending_sptr = NULL;
+        ucs_status_ptr_t pending_sptr = NULL;
+        req_counter = 0;
+        ucp::data_type_desc_t sdt_desc(m_dt, sbuf.ptr(), size);                               
+        ucp_mem_h memh;
+
+        if (prereg()) {
+            memh = sender().mem_map(sbuf.ptr(), size);
+        }
+
+        set_am_data_handler(receiver(), TEST_AM_NBX_ID, am_data_cb, this,
+                                    data_cb_flags);
+
+        pending_sptr = fill_sq(sdt_desc, memh, header_size, get_send_flag() | flags, data_cb_flags);
+
+        if (pending_sptr == send_completed) {
+            UCS_TEST_MESSAGE <<"Failed to get pending request";
+        }
+
+        rndv_pending_sptr = send_am(sdt_desc, get_send_flag() | flags,
+                                            m_hdr_copy.data(), header_size*2, memh);
+
+        ucs::fill_random(m_hdr_copy);
+
+        while(progress());
+        wait_for_value(&cb_counter, req_counter);
+        request_wait(pending_sptr);
+        wait_for_value(&cb_counter, req_counter);
+        request_wait(rndv_pending_sptr);
+
+        EXPECT_EQ(req_counter, cb_counter);
+        if (prereg()) {
+            sender().mem_unmap(memh);
+        }
+    }
+
+    static ucs_status_t am_data_cb(void *arg, const void *header,
+                                   size_t header_length, void *data,
+                                   size_t length,
+                                   const ucp_am_recv_param_t *param)
+    {   
+        test_ucp_am_nbx_send_copy_header *self = reinterpret_cast<test_ucp_am_nbx_send_copy_header*>(arg);
```

Comment:
Unresolved

Reply:
Will be fixed.
I see uses of this cast in many places in this file... for example:
test_ucp_am_nbx *self = reinterpret_cast<test_ucp_am_nbx*>(arg);
And many more...

Reply:
Yes, we sometimes have "not ideal" code (like with the "UCS deadline" you had mentioned before).
It might be fixed someday, but in a separate commit.